### PR TITLE
Add a list-based stack implementation

### PR DESCRIPTION
**Pull request by muqiuhan**
In this PR I implemented a Stack based on the existing List in the standard library.

### CODE REVIEW

**Comment by muqiuhan** [comment]
The reason why cla-check failed is that before the first commit, I failed to authorize github due to network problems. Now it has been successful:

image

**Comment by Lampese** [comment]
> The reason why cla-check failed is that before the first commit, I failed to authorize github due to network problems. Now it has been successful:
> 
> image
> 

You could update your branch(for example submit a new commit or merge main branch) to refersh the check.

**Comment by muqiuhan** [comment]
> You could update your branch(for example submit a new commit or merge main branch) to refersh the check.

The two commits 1e5165c and ed99e83 were submitted after successfully authorizing github, but they still failed after triggering cla-check.

**Comment by lijunchen** [comment]
Did you click the agree button after logging in? @muqiuhan

**Comment by muqiuhan** [comment]
> Did you click the agree button after logging in? @muqiuhan

Thank you for your reply. After logging in again, cla-check has now passed.

File: stack/stack.mbt [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+  @assertion.assert_true(empty.is_empty())?
+}
+
+/// Returns the number of elements of the Stack
+pub fn length[T](self : Stack[T]) -> Int {
+  self.len
+}
+
+// Compares whether the elements in two lists are the same
+// pub fn equal[T](self: Stack[T], other: Stack[T]) -> Bool {
```
Comment (Yoorkin) [inline stack/stack.mbt]: Did you forget to uncomment this code? The method name of trait `Eq` is `op_equal`

File: stack/stack.mbt [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+// }
+
+/// Iterates over the stack.
+/// 
+/// # Example
+/// 
+/// ```
+/// Stack::from_array([1, 2, 3, 4, 5]).iter(print) // output: 54321
+/// ```
+pub fn iter[T](self : Stack[T], f : (T) -> Unit) -> Unit {
```
Comment (Yoorkin) [inline stack/stack.mbt:435]: I think the methods of the stack shouldn't manipulate the elements except for the top element. If someone needs it, they actually need a `List`.

**Review by Yoorkin — CHANGES_REQUESTED** [review]

File: stack/stack.mbt (hunk 2 above)
Comment (muqiuhan) [reply stack/stack.mbt:435]: I think what you said makes sense. I thought so too at first, but I was just not sure whether I should add it, so I added it.

Comment (muqiuhan) [reply stack/stack.mbt]: I removed these functions in a new commit [569c3eb](https://github.com/moonbitlang/core/pull/40/commits/569c3eb8110044e78d8ad1d87b8ca8392b7efaa5)

Comment (Yoorkin) [reply stack/stack.mbt]: > I removed these functions in a new commit [569c3eb](https://github.com/moonbitlang/core/pull/40/commits/569c3eb8110044e78d8ad1d87b8ca8392b7efaa5)

Nice! Can you solve the conflicts before merging?

File: list/list.mbt [hunk 3]
```diff
@@ -822,3 +822,217 @@ test "lookup" {
...
+
+/// Checks if all elements of the list satisfy the predicate f.
+///
+/// # Example
+///
+/// ```
+/// println(List::[2, 4, 6].forall(fn(element) -> Bool { element % 2 == 0 }))
+/// // output: true
+/// ```
+pub fn forall[T](self : List[T], f : (T) -> Bool) -> Bool {
```
[the remark this answers is not in the document]
Comment (muqiuhan) [reply list/list.mbt]: I'm sorry for my carelessness, I deleted it in the commit [8aa1eec](https://github.com/moonbitlang/core/pull/40/commits/8aa1eecf1c95171cb49537a263022e95d20809ea).

**Pull request merged**